plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs

```diff
@@ -73,230 +151,453 @@ their margin. If either party fails to make a margin payment then the contract
+    l == r = ftsShortMargin l == ftsShortMargin r && ftsLongMargin l == ftsLongMargin r
+
+data Role = Long | Short
+    deriving (Generic, Show)
+
+instance Eq Role where
+    Long == Long = True
+    Short == Short = True
+    _ == _ = False
+
+{-# INLINABLE adjustMargin #-}
+adjustMargin :: Role -> Value -> MarginAccounts -> MarginAccounts
+adjustMargin role value accounts =
+    case role of
+        Long  -> accounts { ftsLongMargin = ftsLongMargin accounts + value }
+        Short -> accounts { ftsShortMargin = ftsShortMargin accounts + value }
+
+{-# INLINABLE totalMargin #-}
+totalMargin :: MarginAccounts -> Value
+totalMargin MarginAccounts{ftsShortMargin, ftsLongMargin} =
+    ftsShortMargin + ftsLongMargin
+
+data FutureState = 
+    Running MarginAccounts
+    | Finished
+    deriving (Haskell.Eq)
+
+instance Eq FutureState where
+    Running ma == Running ma' = ma == ma'
+    Finished   == Finished    = True
+    _ == _ = False
+
+data FutureAction =
+    AdjustMargin Role Value
+    | Settle (OracleValue Value)
+    -- ^ Close the contract at the delivery date by making the agreed payment
+    --   and returning the margin deposits to their owners
+    | SettleEarly (OracleValue Value)
+    -- ^ Close the contract early after a margin payment has been missed. 
+    --   The value of both margin accounts will be paid to the role that
+    --   *didn't* violate the margin requirement
+
+{-# INLINABLE futureStateMachine #-}
+futureStateMachine
+    :: Future
+    -> FutureOwners
+    -> StateMachine FutureState FutureAction
+futureStateMachine ft fos =
+    StateMachine
+        { smTransition = futureTransition
+        , smCheck      = futureCheck ft fos
+        , smFinal      = \case { Finished -> True; _ -> False }
+        }
 
+{-# INLINABLE mkValidator #-}
+mkValidator :: Future -> FutureOwners -> Scripts.ValidatorType (SM.StateMachine FutureState FutureAction)
+mkValidator future ftos = SM.mkValidator (futureStateMachine future ftos)
+
+scriptInstance :: Future -> FutureOwners -> Scripts.ScriptInstance (SM.StateMachine FutureState FutureAction)
+scriptInstance future ftos = 
+    let val = $$(PlutusTx.compile [|| validatorParam ||])
+            `PlutusTx.applyCode`
+                PlutusTx.liftCode future
+                `PlutusTx.applyCode`
+                    PlutusTx.liftCode ftos
+        validatorParam f g = SM.mkValidator (futureStateMachine f g)
+        wrap = Scripts.wrapValidator @FutureState @FutureAction
+    
+    in Scripts.Validator @(SM.StateMachine FutureState FutureAction)
+        val
+        $$(PlutusTx.compile [|| wrap ||])
+        
+machineInstance
+    :: Future
+    -> FutureOwners
+    -> SM.StateMachineInstance FutureState FutureAction
+machineInstance future ftos = 
+    let machine = futureStateMachine future ftos
+        script  = scriptInstance future ftos
+    in SM.StateMachineInstance machine script
+
+validatorScript :: Future -> FutureOwners -> ValidatorScript
```

Comment:
At this point I'm confused: why does this exist? Shouldn't we be using the state machine validator?

Reply:
You're also not using the state machine functions for creating transactions. So I'm just confused as to what degree this is or isn't using the state machine lib.

Reply:
You're right this should be using the state machine validator. I'll fix that.

The functions for creating transactions use the `WalletAPI` class.

Reply:
The state machine lib functions also give you partial transactions. You can still submit them like normal.

Reply:
Possibly we need versions that work with `UnbalancedTx`. So many different ways of making transactions...